### PR TITLE
fix(core): Add stage prefix to stack name shortening process

### LINKED ISSUE #23628: aws-eks: Imported cluster yields invalid long stack name
### Describe the bug

We try adding manifests to an imported cluster, which creates a new stack to host the kubectl provider.
This stack has a name with > 128 characters and fails validation.

As far as I can tell, the problem lies within the [generateStackName](https://github.com/aws/aws-cdk/blob/1485f48c515f81fa585a325e02f1602f56443605/packages/@aws-cdk/core/lib/stack.ts#L1400) method.
The fix to the original [issue](https://github.com/aws/aws-cdk/issues/20124) made changes to the [makeStackName](https://github.com/aws/aws-cdk/pull/20528/files#diff-caa054ee92a558f0c56b9e350128d032b8504f4d30193f70a9a8a4a2e707c990R1318) function and applied the length limit there.
This ensures that the stack id created by generateStackId is no longer than 128 characters.
However, the generated stack id is prefixed with `${assembly.stageName}-` when available, result in a too long stack name

### Expected Behavior

The stack name for the kubectlProvider stack is valid

### Current Behavior

The stack name for the kubectlProvider stack is longer than 128 characters and therefore invalid.

### Reproduction Steps

//Import cluster in a staged application with long names
this.cluster = Cluster.fromClusterAttributes(this, 'ClusterImport', {
      clusterName
      openIdConnectProvider,
      kubectlSecurityGroupId,
      kubectlPrivateSubnetIds,
      kubectlRoleArn,
      vpc,
    });
//Add manifest, resulting in creation of kubectlProvider
    this.cluster.addManifest('SampleNamespace', {
      apiVersion: 'v1',
      kind: 'Namespace',
      metadata: { name: 'SampleNamespace' },
    });

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### CDK CLI Version

2.59.0 (build b24095d)

### Framework Version

_No response_

### Node.js Version

18.13 LTS

### OS

Ventura 13.1

### Language

Typescript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by QustodioPablo**
There is an issue in the stack name generation process where the prefix generated from assembly's stage name is not taken into account when shortening a stack name to meet the requirement of being equal or less than 128 characters longs.

This can lead to generating an invalid stack name greater than 128 characters: since the stack name is shortened to 128 characters, when the prefix is added, the limit is exceeded.

Current solution:
- Add the prefix into the first ID if any prefix is present.

Alternative solution:
- Add the prefix at the start of the ID list. Doing this would require to add more conditionals and consider more corner cases to keep the names consistent with the old algorithm.
 

Fixes #23628

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by QustodioPablo** [comment]
Clarification Request: I was unable to run the integration tests, but created a unit test, and made sure the change was not breaking the already existing unit tests. Do I need to create integration tests for this PR?

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]
We do this in a lot of places, adding prefixes. I'm thinking it's a better options to add prefix as an optional prop in `makeUniqueResourceName` and adding the prefix there. What do you think? Also, I if this isn't impacting the outcome of any integ tests, I think omitting one is fine.

File: packages/@aws-cdk/core/lib/stack.ts [hunk 1]
```diff
@@ -1629,7 +1633,7 @@ export function rootPathTo(construct: IConstruct, ancestor?: IConstruct): IConst
  * behavior.
  */
 function makeStackName(components: string[]) {
-  if (components.length === 1) { return components[0]; }
+  if (components.length === 1 && components[0].length <= 128) { return components[0]; }
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/core/lib/stack.ts:1636]: Iirc, this check is done inside `makeUniqueResourceName` so it shouldn't be needed here.

Comment (QustodioPablo) [reply packages/@aws-cdk/core/lib/stack.ts:1643]: The check done inside `makeUniqueResourceName` will remove the `-` from the prefix, as it removes anything that is not a number or a character, unless you specify to keep a special character though parameters, but I was afraid of breaking compatibility with already existing stacks if doing so.
I'll make changes to pass the prefix to the `makeUniqueResourceName` as an optional parameter as you mentioned in another comment.
Thanks!

File: packages/@aws-cdk/core/lib/private/unique-resource-name.ts [hunk 2]
```diff
@@ -68,6 +68,9 @@ export function makeUniqueResourceName(components: string[], options: MakeUnique
 
   // Calculate the hash from the full path, included unresolved tokens so the hash value is always unique
   const hash = pathHash(components);
+  if (prefix) {
```
Comment (QustodioPablo) [inline packages/@aws-cdk/core/lib/private/unique-resource-name.ts:71]: The prefix is added after the hash calculation to preserver the same hash for already existing stacks that fall within this scenario.
I'm not sure if I'm worrying too much, as I'm not sure what happens if you create a stack with CDK, then after this change you run CDK again, and the old created stack's name does not match the new name generated.

File: packages/@aws-cdk/core/lib/stack.ts [hunk 3]
```diff
@@ -1635,9 +1635,14 @@ export function rootPathTo(construct: IConstruct, ancestor?: IConstruct): IConst
...
  * behavior.
  */
-function makeStackName(components: string[]) {
-  if (components.length === 1) { return components[0]; }
-  return makeUniqueResourceName(components, { maxLength: 128 });
+function makeStackName(components: string[], prefix: string='') {
+  if (components.length === 1) {
+    const stack_name = prefix + components[0];
+    if (stack_name.length <= 128) {
+      return prefix + components[0];
```
Comment (QustodioPablo) [inline packages/@aws-cdk/core/lib/stack.ts:1642]: Where there is a single component and a prefix, passing both the components and the prefix to the `makeUniqueResourceName` might seem like the simplest and most straight forward implementation, but there is an issue with this approach.
Doing so will make some tests fail, where a stage with an empty stack is defined. Doing so will generate a prefix (the stage name), and a single component with the name of the default stack. The `makeUniqueResourceName` removes all components with the `default` name, leaving us with no resources in the list, which makes `makeUniqueResourceName` throw an error.

File: packages/aws-cdk-lib/core/lib/private/unique-resource-name.ts [hunk 4]
```diff
@@ -46,7 +46,7 @@ const MAX_LEN = 256;
 
 const HASH_LEN = 8;
 
-export function makeUniqueResourceName(components: string[], options: MakeUniqueResourceNameOptions) {
+export function makeUniqueResourceName(components: string[], options: MakeUniqueResourceNameOptions, prefix: string='') {
```
Comment (TheRealAmazonKendra) [inline packages/aws-cdk-lib/core/lib/private/unique-resource-name.ts:49]: Let's add this as an option instead independently like this.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Totally understand your concerns about these changes causing breaking changes, but this is adding a lot of extra handling to avoid that. I'm thinking that it might be best to put these changes behind a feature flag so that the changes can be made without worrying about that. What do you think?

File: packages/aws-cdk-lib/core/lib/private/unique-resource-name.ts (hunk 4 above)
Comment (QustodioPablo) [reply packages/aws-cdk-lib/core/lib/private/unique-resource-name.ts]: Done!

**Comment by QustodioPablo** [comment]
> Totally understand your concerns about these changes causing breaking changes, but this is adding a lot of extra handling to avoid that. I'm thinking that it might be best to put these changes behind a feature flag so that the changes can be made without worrying about that. What do you think?

I'm ok with adding this behind a feature flag, but we have stacks deployed with an older version that needs updating. We were not able to upgrade our CDK version yet because of this issue, so I'd like to try and fix this issue while keeping it compatible with older versions as much as possible, since afaik it would be mean having to redeploy everything in case the naming changes.

Any suggestion?

**Comment by QustodioPablo** [comment]
Gah, I was away when the PR was approved and merged, and didnt have a chance to implement the feature flag.
Will look into it and create a new PR when done.

**Pull request merged**